[Dependent on PR-340] OSMT-301: Added ability to set multiple authors for an RSD

trs-aaron opened the pull request:
Resolves https://github.com/wgu-opensource/osmt/issues/301

Depends on https://github.com/wgu-opensource/osmt/pull/340

Created RichSkillAuthors table and added ability to set multiple authors on an RSD.

[![Java CI with Maven](https://github.com/wgu-opensource/osmt/actions/workflows/maven.yml/badge.svg?branch=feature%2Fosmt-301_shared-authorship)](https://github.com/wgu-opensource/osmt/actions/workflows/maven.yml)

JohnKallies commented on api/src/main/kotlin/edu/wgu/osmt/richskill/RichSkillDescriptorTable.kt:
```diff
@@ -29,12 +29,37 @@ object RichSkillDescriptorTable : LongIdTable("RichSkillDescriptor"), TableWithU
-    val author = reference(
+}
+
+object RichSkillAuthors : Table("RichSkillAuthors") {
```
I thought we agreed to extending the reuse of the RichSkillKeywords table, since Author is already implemented as a Keyword (Named Reference). Would you please help me understand the value in adding an extra table for this?

JohnKallies replied:
`RickSkillKeywords` is how we have done it with all of the other named references. I see value in keeping that consistency. We can hide the `where` implementations with a few enums used as parameters. I would prefer that we not add a new table that is functionally redundant. If you see a compelling need, we can chat or speak in Slack? Please advise?

JohnKallies commented:
@trs-aaron and @ottonomy , this change brings a change to our OSMT API. The OpenAPI spec file should be updated as well.

We also need to document the impact on existing OSMT instances that deploy this. Is it a new API version? Will this break 3rd party integrations? To clarify, we did agree that WGU would do the house keeping on the older API files.

JohnKallies commented on api/src/main/kotlin/edu/wgu/osmt/csv/BatchImportRichSkill.kt:
```diff
@@ -175,7 +175,7 @@ class BatchImportRichSkill: CsvImport<RichSkillRow> {
                     collections = collections?.let {ListFieldUpdate(add = it)},
                     jobCodes = allJobcodes?.let { ListFieldUpdate(add = it) },
-                    author = NullableFieldUpdate(keywordRepository.getDefaultAuthor())
+                    authors = ListFieldUpdate(add = listOf(keywordRepository.getDefaultAuthor()))
```
When I started reviewing, I wondered if this would be updated as well. I appreciate the completeness! :)

trs-aaron replied:
> @trs-aaron and @ottonomy , this change brings a change to our OSMT API. The OpenAPI spec file should be updated as well.
> 
> We also need to document the impact on existing OSMT instances that deploy this. Is it a new API version? Will this break 3rd party integrations? To clarify, we did agree that WGU would do the house keeping on the older API files.

@JohnKallies I am doing this and setting to new major version as we discussed. I wasn't sure when to commit this since there is the change to `categories` as well. What I planned was to create a new branch with the OpenAPI & CSV specs with both changes. Then you all could also add changes from the filters work, and any other changes planned, and release the specs all at once.

Meant to have that done last night, but I didn't feel well yesterday afternoon. Its what I'm working on this morning.

manuel-delvillar commented:
In my local migrations fails because I have rows in RichSkillDescriptor with author_id null.
Also seems that before this change author_id can be null and now is not null.
is this only a problem with my data or we should to consider author_id can be null in some cases?

JohnKallies commented:
Based on the filters work that was recently merged (fb6f0297 / https://github.com/wgu-opensource/osmt/pull/312), the multiple authors does not filter correctly. We added a filter state to the ApiSearch object that carries a set of optional List<String> arguments for filter state.

@trs-aaron would you please ensure that multiple authors works correctly with this newer feature? Thank you

trs-aaron commented:
> @trs-aaron and @ottonomy , this change brings a change to our OSMT API. The OpenAPI spec file should be updated as well.
> 
> We also need to document the impact on existing OSMT instances that deploy this. Is it a new API version? Will this break 3rd party integrations? To clarify, we did agree that WGU would do the house keeping on the older API files.

@JohnKallies I am doing this and setting to new major version as we discussed. I wasn't sure when to commit this since there is the change to `categories` as well. What I planned was to create a new branch with the OpenAPI & CSV specs with both changes. Then you all could also add changes from the filters work, and any other changes planned, and release the specs all at once.

Meant to have that done last night, but I didn't feel well yesterday afternoon. Its what I'm working on this morning.

JohnKallies commented:
@trs-aaron Based on internal WGU momentum, we are going to be working around our OpenAPI file anyways.

Don't worry about combining PR branches... The next release coming off develop will have an audited and well-curated OpenAPI spec and a major 3.0.0 release of OSMT code / artifacts.

For this work, just please make the updates to the file that reflect the enhancements your team is doing.

trs-aaron commented:
> Based on the filters work that was recently merged ([fb6f029](https://github.com/wgu-opensource/osmt/commit/fb6f0297caafee6cb2a432c38b25ecc674de1b49) / #312), the multiple authors does not filter correctly. We added a filter state to the ApiSearch object that carries a set of optional List arguments for filter state.
> 
> @trs-aaron would you please ensure that multiple authors works correctly with this newer feature? Thank you

I refactored that in at the last and checked that it worked on RSD & Collection lists and it seemed to work fine. I'll go check on Advanced Search.

JohnKallies commented:
We have an inconsistency in the column names when batch importing using the default CSV template. That template exists in a C-Sky NPM module. That's probably not in scope to update right now. Any ideas?

ottonomy commented on api/src/test/kotlin/edu/wgu/osmt/richskill/RichSkillEsRepoTest.kt:
```diff
@@ -375,11 +375,12 @@ class RichSkillEsRepoTest @Autowired constructor(
     @Test
     fun testByApiSearchWithAdvance() {
         // Arrange
+        val authorsList = listOf("author")
```
Good consideration. I'm not sure that this one should be in scope for this issue. I'd love to identify whether author is the only scenario where this type of keyword-based searching would cause false positives due to search finding responsive items where terms appeared across multiple distinct values.

ottonomy commented:
Thanks for getting eyes on this and kickstarting some discussion.

Summary of where we're at right now:
* I see that the approach we had discussed in our meeting about creating a new join table for authors missed the caveat that there are multiple types of keywords already with type-specific filtering implemented across different keywords joined with `RichSkillKeywords`, so no new join table is needed. Generally it is fine to have multiple join tables for different purposes, but given that there's already a pattern to use `keyword.type` for differentiation of different types of joins, it makes sense to stay consistent. **Action:** @trs-aaron will refactor this branch to avoid creation of a new table.
* Update the template spreadsheet `/ui/src/assets/osmt-batch-import-example.csv` to account for field name change (@JohnKallies I didn't see where a default CSV was in a separate CSky npm package.. the only such package I see is the design pattern library, which I agree should be refactored but not in the scope at the moment). I think the example CSV checked into the osmt repo can be updated though. **Action** @trs-aaron please update this CSV as needed.
* Request to update tests and mock data to ensure search response doesn't return results across multiple authors with partial matches of search term: I think no action should be taken on this one for the time being. False positives on search results are not as bad as false negatives, but this could be a nice tiny tiny improvement to make at some point in the future if it proves problematic. 
* We should double check query sent to Elasticsearch as noted [here](https://github.com/wgu-opensource/osmt/pull/326#discussion_r1132673773) **Action**: @trs-aaron to demonstrate successful return of search results based on authors input via tests and demo.

JohnKallies commented:
> 

Yeah, I just assumed it was in the NPM because it wound up in the ui folder in the Spring app, my error. I agree with the approach and the request.

@ottonomy I'm also OK with the whole of your remark.

trs-aaron commented:
@JohnKallies, @jchavez137

Code has been refactored to use existing keywords bridge table. This also uses refactored Search Select Fields so Authors being required in UI is handled correctly. Because this branch is based on that branch that commit shows in this PR as well. Please just focus on this commit [OSMT-301: Added ability to set multiple authors for an RSD](https://github.com/wgu-opensource/osmt/pull/326/commits/1ddcf43536f8367c320e5dc8686ad6f42971168e).

The ES query was not altered. Search seems to be working as you would expect and query is uniform with other Keyword base fields.

trs-aaron commented on api/src/main/kotlin/edu/wgu/osmt/richskill/RichSkillEsRepo.kt:
```diff
@@ -231,7 +231,7 @@ class CustomRichSkillQueriesImpl @Autowired constructor(override val elasticSear
             }
             authors?. let {
-                bq.must(buildNestedQueries(RichSkillDoc::author.name, it))
+                bq.must(buildNestedQueries(RichSkillDoc::authors.name, it))
```
Changing this to `bq.must(generateTermsSetQueryBuilder(RichSkillDoc::authors.name, it))` causes many of the Filtered Search tests to fail. Those tests expect authors query to be handled as an `OR` but if i'm reading this right I think that `generateTermsSetQueryBuilder` creates an implicit `AND` with `.setMinimumShouldMatchScript(Script(list.size.toString()))`. 

Also, other Keywords treat it as an ` AND `  so I'm not sure what expected functionality would be.

Corpratespaz reviewed (changes requested):
- [ ] Update KeywordTypeEnum to 'Authors'

https://github.com/wgu-opensource/osmt/blob/feature/osmt-301_shared-authorship/api/src/main/kotlin/edu/wgu/osmt/keyword/KeywordTypeEnum.kt#L15

trs-aaron commented:
> * [ ]  Update KeywordTypeEnum to 'Authors'
> 
> https://github.com/wgu-opensource/osmt/blob/feature/osmt-301_shared-authorship/api/src/main/kotlin/edu/wgu/osmt/keyword/KeywordTypeEnum.kt#L15

This is inconsistent with the established naming convention as all others are singular. Nothing has changed about the records that are being stored in the Keyword table, they are still individual authors so the singular `KeywordTypeEnum.Author` is still appropriate. The only thing that changes is now multiple Author keyword records can be related to the same RSD.

This would also require the migration of all current records in the `Keyword` table they enum value is stored in `keyword_type_enum` table.

jchavez137 commented on api/src/main/kotlin/edu/wgu/osmt/richskill/RichSkillRepository.kt:
```diff
@@ -285,12 +281,22 @@ class RichSkillRepositoryImpl @Autowired constructor(
         skillUpdate.alignments?.let { lookupAlignments(it, KeywordTypeEnum.Alignment) }
         skillUpdate.employers?.let { lookupReferences(it, KeywordTypeEnum.Employer) }
 
+        val authorsUpdate = if (addingAuthors.size > 0 || removingAuthors.size > 0) {
```
Where are we using this, looks like it should be used between line 319 and line 331, but I think the authors changes are included in the allKeyWordsUpdate object. So my guess is that this is not needed.

trs-aaron replied:
This was removed.

ottonomy reviewed (approved):
To expedite, I see that all the requested changes made by the goal time have been made. Thanks!
